File: src/rapaio/core/stat/ROC.java
/*
 * Apache License
 * Version 2.0, January 2004
 * http://www.apache.org/licenses/
 *
 *    Copyright 2013 Aurelian Tutuianu
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package rapaio.core.stat;

import rapaio.core.Printable;
import rapaio.data.*;

import static rapaio.data.filters.BaseFilters.sort;


/**
 * User: Aurelian Tutuianu <paderati@yahoo.com>
 */
public class ROC implements Printable {

    private final Var score;
    private final Var classes;
    private Frame data;
    private double auc;

    public ROC(Var score, Var actual, Var predict) {
        this.score = score;
        this.classes = Index.newEmpty(actual.rowCount());
        for (int i = 0; i < actual.rowCount(); i++) {
            if (actual.label(i).equals(predict.label(i))) {
                classes.setIndex(i, 1);
            } else {
                classes.setIndex(i, 0);
            }
        }
        compute();
    }

    public ROC(Var score, Var actual, int index) {
        this(score, actual, actual.dictionary()[index]);
    }

    public ROC(Var score, Var actual, String label) {
        this.score = score;
        this.classes = Index.newEmpty(actual.rowCount());
        for (int i = 0; i < actual.rowCount(); i++) {
            if (actual.label(i).equals(label)) {
                classes.setIndex(i, 1);
            } else {
                classes.setIndex(i, 0);
            }
        }
        compute();
    }

    private void compute() {
        int p = 0;
        int n = 0;
        double prevtp = 0;
        double prevfp = 0;
        auc = 0;
        for (int i = 0; i < classes.rowCount(); i++) {
            if (classes.missing(i)) {
                continue;
            }
            if (classes.index(i) > 0) p++;
            else n++;
        }


        double fp = 0;
        double tp = 0;
        auc = 0;

        Var rows = sort(Index.newSeq(score.rowCount()), RowComparators.numericComparator(score, false));
        int len = 1;
        double prev = Double.MIN_VALUE;
        for (int i = 0; i < rows.rowCount(); i++) {
            if (score.missing(rows.index(i)) || classes.missing(rows.index(i))) continue;
            if (score.value(rows.index(i)) != prev) {
                prev = score.value(rows.index(i));
                len++;
            }
        }
        data = Frames.newMatrix(len, "threshold", "fpr", "tpr", "acc");
        prev = Double.POSITIVE_INFINITY;
        int pos = 0;

        for (int i = 0; i < rows.rowCount(); i++) {
            if (score.missing(rows.index(i)) || classes.missing(rows.index(i))) continue;
            if (score.value(rows.index(i)) != prev) {
                auc += Math.abs(prevfp - fp) * Math.abs(prevtp + tp) / 2.;
                double accValue = (tp + n - fp) / (0. + n + p);
                data.setValue(pos, "threshold", prev);
                data.setValue(pos, "fpr", fp / (1. * n));
                data.setValue(pos, "tpr", tp / (1. * p));
                data.setValue(pos, "acc", accValue);
                prevfp = fp;
                prevtp = tp;
                prev = score.value(rows.index(i));
                pos++;
            }
            if (classes.index(rows.index(i)) > 0) tp++;
            else fp++;
        }
        data.setValue(pos, "threshold", prev);
        data.setValue(pos, "fpr", 1.);
        data.setValue(pos, "tpr", 1.);
        data.setValue(pos, "acc", p / (0. + n + p));

        auc += Math.abs(n - prevfp) * (p + prevtp) / 2.;
        auc /= (1. * p * n);
    }

    public Frame getData() {
        return data;
    }

    public double auc() {
        return auc;
    }

    @Override
    public void buildSummary(StringBuilder sb) {
        sb.append("> ROC summary\nnot implemented\n");
    }
}

File: src/rapaio/core/stat/ROC.scala
/*
 * Apache License
 * Version 2.0, January 2004
 * http://www.apache.org/licenses/
 *
 *    Copyright 2013 Aurelian Tutuianu
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package rapaio.core.stat

import rapaio.data.{Index, Feature, Frame}
import rapaio.printer.Summarizable

/**
 * User: Aurelian Tutuianu <paderati@yahoo.com>
 */
@deprecated("not fully implemented", "1.0")
class ROC extends Summarizable {
  private var score: Feature = null
  private var classes: Feature = null
  private var _data: Frame = null
  private var _auc: Double = 0.0

  private def compute(): Unit = {
    var p: Int = 0
    var n: Int = 0
    var prevtp: Double = 0
    var prevfp: Double = 0
    _auc = 0.0
    for (i <- 0 until classes.rowCount) {
      if (!classes.missing(i)) {
        if (classes.indexes(i) > 0) p += 1
        else n += 1
      }
    }

    var fp: Double = 0
    var tp: Double = 0
    _auc = 0
    //  val sort: Feature = sort(score, RowComparators.numericComparator(score, false))
    val sort: Feature = null
    var len: Int = 1
    var prev: Double = Double.MinValue
    for (i <- 0 until sort.rowCount) {
      if (!(sort.missing(i) || classes.missing(sort.rowId(i)))) {
        if (sort.values(i) != prev) {
          prev = sort.values(i)
          len += 1
        }
      }
    }
    //  data = Frames.newMatrixFrame(len, Array[String]("threshold", "fpr", "tpr", "acc"))
    // TODO newMatrixFrame
    _data = null
    prev = Double.PositiveInfinity
    var pos: Int = 0
    for (i <- 0 until sort.rowCount) {
      if (!(sort.missing(i) || classes.missing(sort.rowId(i)))) {
        if (sort.values(i) != prev) {
          _auc += math.abs(prevfp - fp) * math.abs(prevtp + tp) / 2.0
          val accValue: Double = (tp + n - fp) / (0.0 + n + p)
          _data.values(pos, "threshold") = prev
          _data.values(pos, "fpr") = fp / (1.0 * n)
          _data.values(pos, "tpr") = tp / (1.0 * p)
          _data.values(pos, "acc") = accValue
          prevfp = fp
          prevtp = tp
          prev = sort.values(i)
          pos += 1
        }
        if (classes.indexes(sort.rowId(i)) > 0) tp += 1
        else fp += 1
      }
    }

    _data.values(pos, "threshold") = prev
    _data.values(pos, "fpr") = 1.0
    _data.values(pos, "tpr") = 1.0
    _data.values(pos, "acc") = p / (0.0 + n + p)
    _auc += math.abs(n - prevfp) * (p + prevtp) / 2.0
    _auc /= (1.0 * p * n)
  }

  def data: Frame = _data

  def getAuc: Double = _auc

  override def buildSummary(sb: StringBuilder): Unit = ???

}

object ROC {

  def apply(score: Feature, actual: Feature, predict: Feature): ROC = {
    val roc = new ROC
    roc.score = score
    roc.classes = new Index(actual.rowCount, actual.rowCount, 0)
    for (i <- 0 until actual.rowCount) {
      if (actual.labels(i) == predict.labels(i)) {
        roc.classes.indexes(i) = 1
      }
      else {
        roc.classes.indexes(i) = 0
      }
    }
    roc.compute()
    roc
  }

  def apply(score: Feature, actual: Feature, index: Int): ROC = {
    apply(score, actual, actual.labels.dictionary(index))
  }

  def apply(score: Feature, actual: Feature, label: String): ROC = {
    val roc = new ROC()
    roc.score = score
    roc.classes = new Index(actual.rowCount, actual.rowCount, 0)
    for (i <- 0 until actual.rowCount) {
      if (actual.labels(i) == label) {
        roc.classes.indexes(i) = 1
      }
      else {
        roc.classes.indexes(i) = 0
      }
    }
    roc.compute()
    roc
  }


}